Add managed MachO signing

jtschuster opened the pull request:
Moving https://github.com/dotnet/runtime/pull/107378 to a new PR since it now has little shared history.

This signer is byte-for-byte identical to the output of `codesign` except for padding at the end of the file and the corresponding size fields in the headers.

The MachObjectFile and CodeSignature store all the relevant headers and signature information to add/remove a signature from a Mach Object. Most other files are details regarding header format and reading files with different endianness.

The new signer uses a MemoryMappedViewAccessor instead of a Stream. Since memory mapped files can't be resized, `MachObjectFile.CreateAdHocSignature()` will write out the headers and load commands to the memory mapped file, but not the signature blob. The signature blob will instead be written to the destination file stream after the rest of the file is copied from the memory mapped file.

The signature is composed of an EmbeddedSignature superblob with an ID and offset of 3 following blobs, followed by a CodeDirectory blob, an empty Requirements blob, and an empty CMS blob. The empty Requirements blob and an empty CMS blob are not strictly required but are added for compatibility with `codesign`. The CodeDirectory includes the identifier string, hashes of each page of the file and information required to verify the hashes.

The signing process for an unsigned executable extends the __LINKEDIT segment (which is always at the end of the file) to make room for the signature blob, writes the signature blob to that location, and adds a CodeSignature load command that points the the signature. If the executable was already signed, it simply clears the old signature and replaces it, extends or shrinks the existing __LINKEDIT size and updates the existing CodeSignature load command.

jtschuster committed:
Add managed MachO signing

jtschuster committed:
Add managed MachO signing

am11 commented on src/installer/managed/Microsoft.NET.HostModel/MachO/CodeSignature.cs:
```diff
@@ -0,0 +1,124 @@
+    private (EmbeddedSignatureHeader Header, long ptr) embeddedSignature;
+    private (CodeDirectoryHeader Header, long ptr) codeDirectory;
+    private (byte[] identifier, long ptr) identifierPtr;
+    private (byte[], long ptr) cdHashes;
```
nit, can give first item a name (better than Item1 😅)
```suggestion
    private (byte[] blob, long ptr) cdHashes;
```

am11 commented on src/installer/managed/Microsoft.NET.HostModel/MachO/CodeSignature.cs:
```diff
@@ -0,0 +1,124 @@
+
+internal class CodeSignature
+{
+    private (EmbeddedSignatureHeader Header, long ptr) embeddedSignature;
```
style nit: _ prefix for private/private readonly fields.

am11 commented on src/installer/managed/Microsoft.NET.HostModel/Microsoft.NET.HostModel.csproj:
```diff
@@ -35,5 +36,12 @@
+      PrivateAssets="all"
+      OutputItemType="Analyzer"
+      ReferenceOutputAssembly="false" /> -->
+  </ItemGroup>
```
I assume it's a `TODO: cleanup before merging`? ;)

jtschuster replied:
Oops, amended the commit and forgot to push before opening the PR. Thanks for pointing that out.

am11 commented on src/installer/managed/Microsoft.NET.HostModel/MachO/EmbeddedSignatureHeader.cs:
```diff
@@ -0,0 +1,33 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.InteropServices;
+
+namespace Microsoft.NET.HostModel.MachO;
+
+[StructLayout(LayoutKind.Sequential)]
```
nit: maybe remove all `StructLayout`? It's default layout, there is no Pack attribute specified or no P/Invoke?
```suggestion
namespace Microsoft.NET.HostModel.MachO;

```

jtschuster replied:
Removing the `StructLayout` causes 'unused field' warnings. I think the StructLayout makes more sense rather than suppressions and makes it clear that these are representations of unmanaged data, but am fine to change it if there's a better solution.

jtschuster committed:
Merge branch 'MachSigning' of https://github.com/jtschuster/runtime into MachSigning

jtschuster committed:
PR feedback, use different memory map file workaround

- Rename (header, offset) tuples
- Prefix underscore on private fields
- Modify load commands in-place in memory mapped file, and write the
  signature to the output stream since memory mapped file can't grow.

am11 commented on src/installer/tests/Microsoft.NET.HostModel.Tests/MachObjectSigning/Data/a.out:
We can add these binary fixtures to https://github.com/dotnet/runtime-assets and create `Microsoft.NET.HostModel.TestData` which can then be consumed as nuget package. Checking in binaries often gets flagged later.

am11 commented on src/installer/tests/Microsoft.NET.HostModel.Tests/MachObjectSigning/Data/a.out:
We can add these binary fixtures to https://github.com/dotnet/runtime-assets/tree/main/src, e.g. `Microsoft.NET.HostModel.TestData` which can then be consumed as nuget package. Checking in binaries often get flagged later.

am11 commented:
`HostActivation.Tests.SelfContainedAppLaunch.RelativeEmbeddedPath` is failing on osx-x64:

> System.ComponentModel.Win32Exception : An error occurred trying to start process '/Users/runner/work/1/s/artifacts/tests/host/osx.x64.Release/ha/4ishikta.mrd/HelloWorld/sub/HelloWorld' with working directory '/Users/runner/work/1/s/artifacts/bin/HostActivation.Tests/Release/net9.0'. Exec format error

jtschuster committed:
Set enableMacOSCodesign=true in singlefile test

am11 commented on src/installer/managed/Microsoft.NET.HostModel/MachO/NameBuffer.cs:
```diff
@@ -0,0 +1,22 @@
+    private ulong _nameLower;
+    private ulong _nameUpper;
+
+    public static NameBuffer __TEXT = BitConverter.IsLittleEndian ?
+        new NameBuffer { _nameLower = 0x0000545845545F5F, _nameUpper = 0x0000000000000000 }
+        : new NameBuffer { _nameLower = 0x5F5F544558540000, _nameUpper = 0x0000000000000000 };
+
+    public static NameBuffer __LINKEDIT = BitConverter.IsLittleEndian ?
+        new NameBuffer { _nameLower = 0x44454B4E494C5F5F, _nameUpper = 0x0000000000005449 }
+        : new NameBuffer { _nameLower = 0x5F5F4C494E4B4544, _nameUpper = 0x4954000000000000 };
```
nit: slightly more readable
```suggestion
    private NameBuffer(ReadOnlySpan<byte> nameBytes)
    {
        byte[] buffer = new byte[16];
        nameBytes.CopyTo(buffer);

        if (BitConverter.IsLittleEndian)
        {
            _nameLower = BitConverter.ToUInt64(buffer, 0);
            _nameUpper = BitConverter.ToUInt64(buffer, 8);
        }
        else
        {
            _nameLower = BitConverter.ToUInt64(buffer, 8);
            _nameUpper = BitConverter.ToUInt64(buffer, 0);
        }
    }

    public static NameBuffer __TEXT = new NameBuffer("__TEXT"u8);
    public static NameBuffer __LINKEDIT = new NameBuffer("__LINKEDIT"u8);
```

am11 commented on src/installer/managed/Microsoft.NET.HostModel/MachO/MachMagicExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+                => value,
+            MachMagic.MachHeader64OppositeEndian or MachMagic.MachHeaderOppositeEndian
+                => BinaryPrimitives.ReverseEndianness(value),
+            _ => throw new InvalidDataException("Invalid magic value")
```
```suggestion
            _ => throw new InvalidDataException($"Invalid magic value 0x{magic:X}")
```

am11 commented on src/installer/managed/Microsoft.NET.HostModel/MachO/MachMagicExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+                => value,
+            MachMagic.MachHeader64OppositeEndian or MachMagic.MachHeaderOppositeEndian
+                => BinaryPrimitives.ReverseEndianness(value),
+            _ => throw new InvalidDataException("Invalid magic value")
```
```suggestion
            _ => throw new InvalidDataException($"Invalid magic value 0x{magic:X}")
```

huoyaoyuan commented on src/installer/managed/Microsoft.NET.HostModel/MachO/RequirementsBlob.cs:
```diff
@@ -0,0 +1,31 @@
+
+    public byte[] GetBytes()
+    {
+        byte[] buffer = new byte[12];
+        BitConverter.GetBytes((uint)_magic).CopyTo(buffer, 0);
+        BitConverter.GetBytes(_requirementsBlobSize).CopyTo(buffer, 4);
+        BitConverter.GetBytes(_entitlementsBlobLength).CopyTo(buffer, 8);
```
Nit: cast the span and write. Array allocation should satisfy the alignment for uint32.

jtschuster replied:
Not sure I follow what you mean here, could you elaborate?

huoyaoyuan replied:
`BinaryPrimitives.WriteInt32BigEndian` etc, which should be available for netstandard2.0 . `BitConverte` is endianness-dependent, so I'm not sure which way is correct.

huoyaoyuan commented on src/installer/managed/Microsoft.NET.HostModel/MachO/CodeSignature.cs:
```diff
@@ -0,0 +1,155 @@
+
+    internal void WriteToStream(FileStream file)
+    {
+        byte[] arr = new byte[(int)Size];
+        Span<byte> buffer = arr;
```
`stackalloc` since it's a constant size?

am11 replied:
This is netstandard2.0 so most of the new stuff doesn't work. In this case, byte[] arr is needed for line 152 (File.Write).

jtschuster committed:
Make NameBuffer constants more readable

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

jtschuster committed:
Update src/installer/managed/Microsoft.NET.HostModel/MachO/MachMagicExtensions.cs

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

jtschuster committed:
Update src/installer/managed/Microsoft.NET.HostModel/MachO/MachMagicExtensions.cs

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

jtschuster commented on src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs:
```diff
@@ -346,33 +347,16 @@ public string GenerateBundle(IReadOnlyList<FileSpec> fileSpecs)
+            if (_macosCodesign && RuntimeInformation.IsOSPlatform(OSPlatform.OSX) && Codesign.IsAvailable)
             {
-                var (exitCode, stdErr) = HostModelUtils.RunCodesign("-s -", bundlePath);
+                var (exitCode, stdErr) = Codesign.Run("-s -", bundlePath);
```
The bundler did something that didn't work well with the other signing implementation, so I left it to use `codesign`. I think it would work with this, but just haven't tried it and was going to do that in a follow up. I can try it and add it to this PR if you'd prefer, though.

jtschuster commented on src/installer/tests/TestUtils/TestApp.cs:
```diff
@@ -93,7 +93,8 @@ private void CreateAppHost(string hostSourcePath, bool isWindowsGui, bool copyRe
                 disableCetCompat: disableCetCompat,
-                dotNetSearchOptions: dotNetRootOptions);
+                dotNetSearchOptions: dotNetRootOptions,
+                enableMacOSCodeSign: RuntimeInformation.IsOSPlatform(OSPlatform.OSX));
```
Yes, I believe `CreateAppHost.DoesNotCodeSignAppHostByDefault` does this.

jtschuster replied:
Ah, you're right, I'll add tests that for without signing

jtschuster commented on src/installer/managed/Microsoft.NET.HostModel/MachO/CodeDirectoryHeader.cs:
```diff
@@ -0,0 +1,121 @@
+/// Version 0x20400
+/// </summary>
+[StructLayout(LayoutKind.Sequential)]
+internal struct CodeDirectoryHeader
```
Yes, I can link to their [CodeDirectory.h](https://github.com/apple-oss-distributions/Security/blob/3dab46a11f45f2ffdbd70e2127cc5a8ce4a1f222/OSX/libsecurity_codesigning/lib/codedirectory.h#L193) here, and find the corresponding types for each of the structs.

agocke commented:
Could you add a short description of what codesigning actually does?

jtschuster committed:
PR Feedback:

- Link to Apple open source code for binary formats
- Remove extra Enum values
- Remove binary test data and use test data package
- Don't add signature to test apps that weren't signed before

jtschuster committed:
Handle folders in the test data package

agocke commented on src/installer/managed/Microsoft.NET.HostModel/MachO/CodeDirectoryHeader.cs:
```diff
@@ -0,0 +1,121 @@
+[StructLayout(LayoutKind.Sequential)]
+internal struct CodeDirectoryHeader
+{
+    private BlobMagic _magic;
```
Reading https://github.com/apple-oss-distributions/Security/blob/3dab46a11f45f2ffdbd70e2127cc5a8ce4a1f222/SecurityTool/sharedTool/codesign.c#L73
it doesn't look like this can be anything other than CSMAGIC_CODEDIRECTORY. Is there a reason to allow people to set anything else?

jtschuster replied:
No real reason, I can change it to readonly with a default value.

agocke commented on src/installer/managed/Microsoft.NET.HostModel/MachO/CodeSignature.cs:
```diff
@@ -0,0 +1,155 @@
+    private (CodeDirectoryHeader Header, long FileOffset) _codeDirectory;
+    private (byte[] Identifier, long FileOffset) _identifierPtr;
+    private (byte[] Blob, long FileOffset) _cdHashes;
+    private (RequirementsBlob Header, long FileOffset) _requirementsBlob;
```
I can't seem to find where this is actually required for code signing. That is, shouldn't the requirements already be present in the binary?

jtschuster replied:
They're present in signed binaries, but not in unsigned ones, and the offset will change if the identifier used is different than the existing signature.

agocke replied:
But this is only for signing our own binaries, right? So we can assume that the incoming binaries haven't been signed?

agocke commented on src/installer/managed/Microsoft.NET.HostModel/MachO/CodeSignature.cs:
```diff
@@ -0,0 +1,155 @@
+        (EmbeddedSignatureHeader Header, long FileOffset) embeddedSignature,
+        (CodeDirectoryHeader Header, long FileOffset) codeDirectory,
+        (byte[] identifier, long FileOffset) identifierPtr,
+        (byte[], long FileOffset) cdHashes,
```
Do we need a separate hashes offset? Don't the hashes appear directly after the identifier? So `identifierPtr.FileOffset + identifier.Length`?

agocke commented on src/installer/managed/Microsoft.NET.HostModel/MachO/CodeSignature.cs:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    public static CodeSignature Create(
+        (EmbeddedSignatureHeader Header, long FileOffset) embeddedSignature,
```
Do we want duplicate FileOffsets at all? Since the signature blob looks contiguous it seems like we need only one file-relative offset -- the offset to the signature blob. Everything else looks like a relative offset.

jtschuster replied:
Yeah, technically we don't need the other pointers, I can remove them.

agocke commented on src/installer/managed/Microsoft.NET.HostModel/MachO/CodeSignature.cs:
```diff
@@ -0,0 +1,155 @@
+        return cs;
+    }
+
+    public static CodeSignature Read(MemoryMappedViewAccessor file, long fileOffset)
```
Do we need this functionality? That is, reading out of a blob? For testing it seems like we could just validate that the bit pattern is correct. We shouldn't necessarily need to rehydrate the whole blob.

jtschuster replied:
We can't just do a byte-for-byte comparison since the hashes for the first page for two will be different if that's what you mean. We might be able to move this method to the test assembly, though.

jtschuster commented on src/installer/managed/Microsoft.NET.HostModel/MachO/CodeDirectoryHeader.cs:
```diff
@@ -0,0 +1,121 @@
+
+    public BlobMagic Magic
+    {
+        get => (BlobMagic)((uint)_magic).ConvertFromBigEndian();
```
I found it easier to keep the memory representation identical to what's in the file so that reading/writing the file is simple. I don't see much of a benefit of changing it personally.

agocke commented on src/installer/managed/Microsoft.NET.HostModel/MachO/MachObjectFile.cs:
```diff
@@ -0,0 +1,500 @@
+    {
+        long commandsPtr = 0;
+        file.Read(commandsPtr, out _header);
+        _nextCommandPtr = ReadCommands(file, in _header, out _codeSignatureLC, out _textSegment, out _textSegment64, out _linkEditSegment, out _linkEditSegment64, out _lowestSection);
```
This is a lot of work to do in a constructor. If you're going to read things out of the file, I would consider inverting the control flow and having a static method that reads things out of the file and then inject the values into the constructor.

agocke commented on src/installer/managed/Microsoft.NET.HostModel/MachO/LinkEditCommand.cs:
```diff
@@ -0,0 +1,28 @@
+
+    public bool IsDefault => this.Equals(default(LinkEditCommand));
+
+    internal uint GetDataOffset(MachHeader header) => header.ConvertValue(_dataOffset);
```
Is this used anywhere? Can it be moved into the testing library?

jtschuster replied:
Yes, we use this when removing a signature, and determining how much of the file needs to be hashed (the signature blob is not included in the code hashes).

agocke commented on src/installer/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs:
```diff
@@ -125,28 +130,42 @@ void RewriteAppHost(MemoryMappedFile mappedFile, MemoryMappedViewAccessor access
+                            else
                             {
-                                MachOUtils.RemoveSignature(fileStream);
+                                if (MachObjectFile.TryRemoveCodesign(memoryMappedViewAccessor, out long? length))
```
Why are we removing codesign if it's already present?

agocke replied:
Oh I see we were doing that below.

agocke commented on src/installer/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs:
```diff
@@ -125,28 +130,42 @@ void RewriteAppHost(MemoryMappedFile mappedFile, MemoryMappedViewAccessor access
+                            BinaryUtils.WriteToStream(memoryMappedViewAccessor, fileStream, appHostLength);
+                            if (!appHostIsPEImage && enableMacOSCodeSign)
+                            {
+                                machObjectFile.WriteCodeSignature(fileStream);
```
Why split these lines up? That is, why not write the code signature above? Or do all the work to generate the signature here?

jtschuster replied:
Since memory mapped files can't be resized, MachObjectFile.CreateAdHocSignature() will write out the headers and load commands to the memory mapped file, but not the signature blob. The signature blob will instead be written to the destination file stream after the rest of the file is copied from the memory mapped file.

agocke replied:
Oh are files larger after codesigning? I thought the signature blob was already in the file. Could we reserve the space in the apphost template to make signing easier?

jtschuster replied:
The platform linker ad-hoc signs the apphost binary when it's built, so there is a signature blob present, but if the file is renamed to a longer name the signature will be slightly larger.

In my first attempt, I tried extending the source apphost and then shrinking it after we write out the destination, but some tests rely on the source apphost being a read-only file, and I think it makes sense to be allow the source to be read-only. I think this also runs into antivirus performance issues. I wasn't able to find a way to make the memory mapped file larger than the file on disk.

We could copy the source apphost to the destination, extend it, and then modify it in place, but my understanding is that modifying the file in place causes antivirus issues.

We could allocate space when we build the apphost binary, but that seems like a significant change just to make this code a bit neater.

For background, the HostWriter uses a `CopyOnWrite` memory mapped accessor to avoid antivirus performance issues, since the modifications to a CopyOnWrite file aren't visible to other processes until they're written to a new file.

jtschuster replied:
I looked into CopyOnWrite and there doesn't seem to be a simple way to edit the file in memory and flush all the changes only when the file is closed to avoid the AV issues.

jtschuster commented on src/installer/managed/Microsoft.NET.HostModel/MachO/CmsBlob.cs:
```diff
@@ -0,0 +1,19 @@
+namespace Microsoft.NET.HostModel.MachO;
+
+[StructLayout(LayoutKind.Sequential)]
+internal struct CmsWrapperBlob
```
I don't think any tools would care about it. This was just to make it as close to identical to `codesign` as possible.

agocke replied:
Got it, so the goal here is byte-for-byte identical to codesign, not the minimum needed to run?

jtschuster replied:
That's right

am11 commented:
@EgorBo, thanks for looking into it. Maybe we can aim to get rid of `USE_HELPERS_FOR_INT_DIV` as part of it? (it's always true on linux-arm so all the `#if !USE_HELPERS_FOR_INT_DIV` type of branches are code rot).

jtschuster committed:
PR feedback

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into MachSigning

agocke commented on src/installer/managed/Microsoft.NET.HostModel/MachO/MachObjectFile.cs:
```diff
@@ -0,0 +1,500 @@
+    }
+
+    /// <summary>
+    /// Clears the old signature and sets the codeSignatureLC to a
```
comment cuts off

jtschuster committed:
Fix test issues

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into MachSigning

agocke reviewed (approved):
Given the goal to bit-for-bit match codesign, this LGTM aside from minor feedback.

If we wanted to relax our requirements to "do the minimum necessary to pass gatekeeper on macos" I think we could delete a fair amount of code. But we can leave that as a potential future option.

jtschuster committed:
Move BinaryFormat files and Enums to separate folders and add notes about endianness of signature data.

jtschuster commented on src/installer/managed/Microsoft.NET.HostModel/MachO/MachObjectFile.cs:
```diff
@@ -0,0 +1,534 @@
+        textSegment64 = default;
+        linkEditSegment = default;
+        linkEditSegment64 = default;
+        long commandsPtr = Marshal.SizeOf<MachHeader>();
```
I went with `Marshal.SizeOf` because `sizeof` required `unsafe` for structs. Would it be better to use `unsafe` rather than `Marshal.SizeOf`?

am11 replied:
`sizeof(MachHeader)` uses `Unsafe.SizeOf<MachHeader>()`, which is JIT intrinsic (and much faster). `Marshal.SizeOf<MachHeader>()` is a QCall (P/Invoke into runtime which is a slower routine).

jtschuster replied:
Updated `MachObjectFile` and `CodeSignature` to be `unsafe` and used `sizeof` instead of `Marshal.Sizeof<>()`.

jtschuster committed:
Extend the AppHost template on disk before modifying and signing. Remove 32 bit segments from MachObjectFile.

jtschuster committed:
Use new MachO types in MachOUtils.cs

jtschuster committed:
Fix comments, use explicit types, update darc dependencies

jtschuster commented on src/installer/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs:
```diff
@@ -127,26 +132,40 @@ void RewriteAppHost(MemoryMappedFile mappedFile, MemoryMappedViewAccessor access
 
-                            if (assemblyToCopyResourcesFrom != null && appHostIsPEImage)
+                        // Save the transformed host.
+                        using (FileStream fileStream = new FileStream(appHostDestinationFilePath, FileMode.Create, FileAccess.ReadWrite))
```
Yeah, I felt it would be helpful to be consistent and explicit with all the different file-related IDisposable objects being opened, some with required access arguments and some without.

am11 commented on src/installer/managed/Microsoft.NET.HostModel/Bundle/Codesign.cs:
```diff
@@ -0,0 +1,38 @@
+{
+    public class Codesign
+    {
+        private const string CodesignPath = @"/usr/bin/codesign";
```
nit:
```suggestion
        private const string CodesignPath = "/usr/bin/codesign";
```

am11 commented on src/installer/managed/Microsoft.NET.HostModel/Bundle/Codesign.cs:
```diff
@@ -0,0 +1,38 @@
+            Debug.Assert(RuntimeInformation.IsOSPlatform(OSPlatform.OSX));
+            Debug.Assert(IsAvailable);
+
+            var psi = new ProcessStartInfo()
+            {
+                Arguments = $"{args} \"{appHostPath}\"",
```
```suggestion
            var psi = new ProcessStartInfo
            {
                Arguments = $"""{args} "{appHostPath}""",
```

am11 replied:
Ah, right that corner case. (space fixes it but that may not serialize correctly to shell args; multi-line defeats the purpose of it being more succulent 🥲)

am11 replied:
Ah, right that corner case. (space fixes it but that may not serialize correctly to shell args; multi-line defeats the purpose of it being more succinct 🥲)

jtschuster committed:
Use unsafe and sizeof instead of Marshal.SizeOf, fix format nits

jtschuster committed:
Rename signatureLC to signatureLoadCommand, remove unnecessary methods

jtschuster committed:
Remove extra 'unsafe' blocks

jtschuster committed:
Revert arguments string change

jtschuster committed:
Re-add ResourceUpdater to HostWriter

jtschuster committed:
Move CodeSignature to a nested type within MachObjectFile, move signature related methods, remove unused structs

jtschuster commented on src/installer/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs:
```diff
@@ -112,54 +113,56 @@ void RewriteAppHost(MemoryMappedFile mappedFile, MemoryMappedViewAccessor access
                             {
-                                using var updater = new ResourceUpdater(fileStream, true);
-                                updater.AddResourcesFromPEImage(assemblyToCopyResourcesFrom);
-                                updater.Update();
```
Yes, that was a mistake. Thank you for catching that.

jtschuster commented on src/installer/managed/Microsoft.NET.HostModel/MachO/MachObjectFile.cs:
```diff
@@ -0,0 +1,534 @@
+        long commandsPtr = 0;
+        file.Read(commandsPtr, out MachHeader header);
+        var nextCommandPtr = ReadCommands(file, in header, out var codeSignatureLC, out var textSegment, out var textSegment64, out var linkEditSegment, out var linkEditSegment64, out var lowestSection);
+        Debug.Assert(linkEditSegment.Command.IsDefault ^ linkEditSegment64.Command.IsDefault);
```
Updated MachObjectFile to expect a 64 bit exe and throw AppHostMachOFormatException of it's not.

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into MachSigning

jtschuster committed:
Set file length after memory mapped file is disposed

am11 commented on src/installer/managed/Microsoft.NET.HostModel/Bundle/Codesign.cs:
What do you think about the following structure?

* **Separate `MachO/` into its own project:** Create a new project called `Microsoft.NET.HostModel.CodeSigning` with a dedicated `.csproj` file and add a `<ProjectReference>` in `Microsoft.NET.HostModel` to reference it.
* **Rename `CodeSign.cs` to `NativeUtility.cs`:** Move `CodeSign.cs` under this new project as `NativeUtility.cs`, using `public static class NativeUtility`, so the API path would become `Microsoft.NET.HostModel.CodeSigning.MachO.NativeUtility.Run(..)`.

This approach would set up a cleaner structure for future improvements by enabling us to:

* Extend the code signing capabilities to additional platforms (or formats), organized alongside `Microsoft.NET.HostModel.CodeSigning/MachO`.
* Reuse this implementation in NativeAOT build integration by allowing cross-platform signing support. For instance, running `dotnet publish -p:PublishAot=true -r osx-arm64` on Windows currently does not include code signing.

jtschuster replied:
One issue with this is I was planning to use the same MachObjectFile class for the SingleFile bundling that MachOUtils.cs does now, but that wouldn't really fit in a "CodeSigning" assembly. We could name it something like "Microsoft.NET.HostModel.NativeBinary", or maybe this could wait for a follow up.

am11 replied:
Sounds good.

If the target is to eventually make this implementation to sign any arbitrary binary, then it could be useful for AOT and other non-corehost based form factors and we can think of suitable structure (in a separate PR). Otherwise, if it is going to be tied to these binaries, then I think it is fine as-is.

jtschuster committed:
Close memory mapped file in test signing method.

jtschuster committed:
PR Feedback

jtschuster commented on src/installer/tests/Microsoft.NET.HostModel.Tests/AppHost/CreateAppHost.cs:
```diff
@@ -447,6 +485,48 @@ private void ResourceWithUnknownLanguage()
 
+        [Fact]
+        [PlatformSpecific(TestPlatforms.OSX)]
+        public void SignedAppHostRuns()
```
Moved to HostActivation.Tests/MachOHostSigningTests.cs.

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into MachSigning

jtschuster committed:
Revert local testing change that enabled codesigning always.

jtschuster committed:
Revert extra change in TestApp.cs

jtschuster committed:
Remove license from ThirdPartyNotices

elinor-fung reviewed (approved):
Thank you! This looks good to me.

Some things for your list of existing follow-ups:
- I think the SDK only enables signing when actually running on macOS - with this, I think we can/should update that default.
- Tests in dotnet/sdk for cross-building for `osx-<arch>` and signing on non-macOS

jtschuster committed:
Add more definitions links, fix test nits.

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into MachSigning

jtschuster commented on src/installer/pkg/THIRD-PARTY-NOTICES.TXT:
```diff
@@ -1449,3 +1449,27 @@ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
+
+License notice for Melanzana MachO signing
```
I don't think this is needed anymore.

jtschuster commented on src/installer/managed/Microsoft.NET.HostModel/MachO/Enums/MachMagic.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace Microsoft.NET.HostModel.MachO;
+
+internal enum MachMagic : uint
```
I added more for the ones I could find.

jtschuster commented on src/installer/tests/Microsoft.NET.HostModel.Tests/AppHost/CreateAppHost.cs:
```diff
@@ -323,40 +345,56 @@ public void DoesNotCodeSignAppHostByDefault()
+            using (TestArtifact artifact = CreateTestDirectory())
+            {
+                string sourceAppHostMock = PrepareAppHostMockFile(artifact.Location);
+                // File.SetAttributes(sourceAppHostMock, FileAttributes.ReadOnly);
```
This should be uncommented.

jtschuster replied:
Test comment. Does this go through if I don't fully submit a review?

jtschuster committed:
Add comment explaining why the file must be written to create signature, move public method above private methods.

jtschuster commented:
/ba-g WASM timeout, sporadic failure.
